jni/src/common/Foreign/JNI/Unsafe.hs

```diff
@@ -1059,3 +1067,44 @@ getDirectBufferCapacity (upcast -> jbuffer) = do
       return capacity
     else
       throwIO DirectBufferFailed
+
+data CleanerAction = DeleteReferences | Terminate deriving Eq
+data GlobalRefCleaner = GlobalRefCleaner
+  { nextBatchToDelete :: IORef [JObject]
+  , nextAction :: MVar CleanerAction
+  , cleanerAsync :: Async ()
+  }
+
+createGlobalRefCleaner :: IO GlobalRefCleaner
+createGlobalRefCleaner = do
+  wakeup <- newEmptyMVar
+  refs <- newIORef []
+  deletingThread <- asyncBound $ runInAttachedThread $ whileM_
+    ((== DeleteReferences) <$> takeMVar wakeup)
+    ( do
+      xs <- atomicModifyIORef refs $ \xs -> ([], xs)
+      mapM deleteGlobalRefNonFinalized xs
+    )
+  return (GlobalRefCleaner refs wakeup deletingThread)
+
+stopGlobalRefCleaner :: GlobalRefCleaner -> IO ()
+stopGlobalRefCleaner cleaner = do
+  putMVar (nextAction cleaner) Terminate
+  wait (cleanerAsync cleaner)
```

Comment:
I think I would just cancel the thread here. Is it necessary to delete the pending global references before terminating the JVM?

Reply:
Also, probably we should use `uninterruptibleMask_` to make this code safe to asynchronous exceptions.